### ISSUE TITLE
Literal dict typed with union drops all union members except the first

### ISSUE THREAD

**Issue by hamdanal** [microsoft/pyright#5887]
**Describe the bug**
See the title.

**Code or Screenshots**

```python
d1: dict[str, str] | dict[int, int] = {}
d2: dict[int, int] | dict[str, str] = {}

reveal_type(d1)
reveal_type(d2)
```
```console
$ pyright t.py
/tmp/t.py
  /tmp/t.py:4:13 - information: Type of "d1" is "dict[str, str]"
  /tmp/t.py:5:13 - information: Type of "d2" is "dict[int, int]"
0 errors, 0 warnings, 2 informations
```

Interestingly, when the dict is assigned to a non-literal, the union types are preserved

```python
d3: dict[str, str] | dict[int, int] = dict()

reveal_type(d3)
```
```console
$ pyright t.py
/tmp/t.py
  /tmp/t.py:3:13 - information: Type of "d3" is "dict[str, str] | dict[int, int]"
0 errors, 0 warnings, 1 information
```

**VS Code extension or command-line**
```console
$ pyright --version
pyright 1.1.325
```

**Comment by erictraut** [comment]
Using normal inference rules, the expression `{}` would be evaluated as type `dict[Unknown, Unknown]`. However, in this case, [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) is used to determine the type of `{}`. 

Since the "expected type" in this case is a union of multiple `dict` types, the type checker must pick one. The current logic picks (somewhat arbitrarily) the first one in the union. It could just as well pick the second one. It needs to pick one because the object constructed via the statement `{}` must have a type at the time it is constructed. It can't be both a `dict[str, str]` and a `dict[int, int]`.

In general, the order in which unions are defined shouldn't determine the evaluated type, so there's a good argument to be made that pyright should enforce some deterministic ordering so the union order doesn't dictate the behavior. I've updated pyright's logic accordingly. With this change in place, the behavior you will see for `d1` and `d2` is consistent.

Regardless of what type is inferred for the expression `{}`, the type of `d1` is [narrowed on assignment](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing) based on the type of the assigned value. You mentioned in the title of the bug that pyright is "dropping all union members except the first". That is what happens when type narrowing is applied, so that is expected behavior.

The case of `d3` in your example above does appear to be a bug. This assignment if `dict()` should narrow the type of `d3` to either `dict[int, int]` or `dict[str, str]`. I'll investigate why type narrowing isn't being applied in this case.

**Comment by erictraut** [comment]
The bug related to the call to `dict()` was not a type narrowing bug but rather a bug in the type evaluation logic for constructors. This will be fixed in the next release. With these fixes in place, you'll see the following (consistent) behavior.

```python
d1: dict[str, str] | dict[int, int] = {}
d2: dict[int, int] | dict[str, str] = {}
d3: dict[str, str] | dict[int, int] = dict()
d4: dict[int, int] | dict[str, str] = dict()

reveal_type(d1)  # dict[int, int]
reveal_type(d2)  # dict[int, int]
reveal_type(d3)  # dict[int, int]
reveal_type(d4)  # dict[int, int]
```

**Comment by hamdanal** [comment]
Thanks for the detailed response and the quick fix. Your reasoning for type narrowing makes sense although I expected that when an explicit annotation is present, it will be used and not further "changed" by the type checker but maybe for the general case, type narrowing based on assignment works better.

I extracted a bigger example of the code that prompted this report with more context and I still find its behavior a slightly unexpected.

```python
def f(x: list[int] | list[str], y: list[int]):
    d: dict[str, str] | dict[int, int] = {}
    reveal_type(d)

    for ix, iy in zip(x, y):
        if isinstance(ix, int):
            d[ix] = iy
        else:
            d[ix] = str(iy)
    reveal_type(d)
    return d

result = f([1], [2])
reveal_type(result)
```

```console
$ pyright t.py
/tmp/t.py
  /tmp/t.py:3:17 - information: Type of "d" is "dict[str, str]"
  /tmp/t.py:7:13 - error: Argument of type "int" cannot be assigned to parameter "__key" of type "str" in function "__setitem__"
    "int" is incompatible with "str" (reportGeneralTypeIssues)
  /tmp/t.py:7:13 - error: Argument of type "int" cannot be assigned to parameter "__value" of type "str" in function "__setitem__"
    "int" is incompatible with "str" (reportGeneralTypeIssues)
  /tmp/t.py:10:17 - information: Type of "d" is "dict[str, str]"
  /tmp/t.py:14:13 - information: Type of "result" is "dict[str, str]"
2 errors, 0 warnings, 3 informations
```

1. I think the reported errors here are false positives. This may be pushing the type checker too much and expecting type narrowing that cannot be realistically done. I'd like to hear from you about this case and how to properly handle it. Of course the real code is a little more complicated but this captures the essence of the case.

2. Should the narrowed type inside the function be propagated to the outside (see `result` in the example). Of course I can add a return type annotation to the function but I am curious because I thought type narrowing applies only to the scope where it is performed. This is just me wondering how all of this works at this point.

**Comment by erictraut** [comment]
I recommend that you read [this documentation](https://microsoft.github.io/pyright/#/type-concepts) and [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced). This will help you understand the core concepts related to static typing, including the notion of type declarations and type narrowing.

> when an explicit annotation is present, it will be used and not further "changed" by the type checker

An annotation is a _type declaration_. It indicates which types are allowed to be assigned to a symbol. A type checker's job is to enforce that invariant. When a type is declared type for a symbol, the type checker doesn't change the declared type, but the type is further narrowed upon assignment and in the presence of type guards. To see why this is important, consider the following:

```python
x: int | str
x = "hi"
reveal_type(x) # Literal["hi"]
print(len(x))
x = 1
reveal_type(x) # Literal[1]
print(x + 1)
```

In this example, the declared type of `x` is `int | str`. This is an invariant. However, the narrowed type of `x` changes throughout the code as new values are assigned to it. This type narrowing allows the type checker to verify that the `len` and `+ 1` operators are valid. Without type narrowing, these operations would be be flagged as invalid because `len` is an illegal operation for an `int`, and `+ 1` is an illegal operation for a `str`.

> I think the reported errors here are false positives

No, these are not false positives. These are legitimate type violations. When you construct a new `dict` instance, its type is fixed at construction time. You can't create an object that is both a `dict[str, str]` and a `dict[int, int]`. The object is one or the other, and the type checker honors that type. When you create a new dict using the `{}` expression, you (or the type checker) need to decide which type it is. Is it a `dict[str, str]` or is it a `dict[int, int]`? Depending on your choice, you can either add str's or int's to it. You cannot add both.

> Should the narrowed type inside the function be propagated to the outside?

No, type narrowing requires code flow analysis, and this applies only within an execution scope (a function or the global scope of a module). Narrowed types do not "leak" outside of an execution scope. However, pyright does use narrowed types to infer function return types if you don't provide an explicit return type annotation. (Mypy doesn't support return type inference, so it requires explicit return types to be specified.)

Static type checking requires you to think a little differently about your code. There may be cases (like this one) where you need to restructure your code slightly to accommodate static type checking.

Here are a few alternatives that type check without error:

```python
def f(x: list[int] | list[str], y: list[int]) -> dict[str, str] | dict[int, int]:
    # Use a dict[Any, Any] here to indicate that you don't want the type
    # checker to enforce the type arguments for the dict. This removes
    # some type safety, but it allows you to leave your existing code
    # largely unmodified.
    d: dict[Any, Any] = {}

    for ix, iy in zip(x, y):
        if isinstance(ix, int):
            d[ix] = iy
        else:
            d[ix] = str(iy)
    return d
```

If you want the type safety, here's an approach that works. Note that I've also added overloaded signatures, which is appropriate for a polymorphic function like this one. I've also switched from `list` to `Sequence` because your code works with any `Sequence`, so it's a more appropriate type than `list`.

```python
from typing import Sequence, overload

@overload
def f(x: Sequence[int], y: Sequence[int]) -> dict[int, int]: ...
@overload
def f(x: Sequence[str], y: Sequence[int]) -> dict[str, str]: ...

def f(x: Sequence[int] | Sequence[str], y: Sequence[int]):
    if len(x) == 0:
        return dict[int, int]()
    elif isinstance(x[0], int):
        return {int(ix): iy for ix, iy in zip(x, y)}
    else:
        return {str(ix): str(iy) for ix, iy in zip(x, y)}
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.326, which I just published. It will also be included in a future release of pylance.

**Comment by hamdanal** [comment]
Thank you for the great explanation and for the fix
